File: src/common/framework/texture.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <GLES/gl.h>
#include "bitmapdata.h"
#include "texture.h"



void textureInit(Texture* tex, BitmapData* data, Rekt sourceRect) {
  float x = sourceRect.x;
  float y = sourceRect.y;
  float w = sourceRect.w;
  float h = sourceRect.h;
  float right = x + w;
  float top =  y + h; 
  GLfloat texCoords[12] = {
    x, top,
    right, top,
    x,  y,
    right, top,
    right,  y,
    x,  y,
  };

  tex->data = data;

  tex->width = w*data->width;
  tex->height = h*data->height;

  memcpy(tex->uvCoords, texCoords, sizeof(GLfloat)*12);
}

File: src/common/framework/texture.nim
when not defined(H_TEXTURE_H): 
  const 
    H_TEXTURE_H* = true
  import 
    rect

  type 
    BitmapData* {.importc: "BitmapData", header: "texture.h".} = object 
    
  type 
    Texture* {.importc: "Texture", header: "texture.h".} = object 
      width* {.importc: "width".}: cint
      height* {.importc: "height".}: cint
      uvCoords* {.importc: "uvCoords".}: array[12, cfloat]
      data* {.importc: "data".}: ptr BitmapData

  proc textureInit*(tex: ptr Texture; data: ptr BitmapData; sourceRect: Rekt) {.
      importc: "textureInit", header: "texture.h".}

